gasprice/gasprice.go

```diff
@@ -0,0 +1,265 @@
+	}
+
+	g.lock.Lock()
+	lastPrice := g.lastPrice
+	lastHeader := g.lastHeaderHash
+	g.lock.Unlock()
+
+	if currentHeader.Hash == lastHeader {
+		// small optimization, if we calculated already the price for given block
+		return new(big.Int).Set(lastPrice), nil
+	}
+
+	var allPrices []*big.Int
+
+	collectPrices := func(block *types.Block) error {
+		baseFee := block.Header.BaseFee
+		txSorter := newTxByEffectiveTipSorter(block.Transactions, baseFee)
+		sort.Sort(txSorter)
+
+		blockMiner := types.BytesToAddress(block.Header.Miner)
+		signer := crypto.NewSigner(g.backend.Config().Forks.At(block.Number()),
+			uint64(g.backend.Config().ChainID))
+		blockTxPrices := make([]*big.Int, 0)
+
+		for _, tx := range txSorter.txs {
+			tip := tx.EffectiveTip(baseFee)
+
+			if tip.Cmp(g.ignorePrice) == -1 {
+				// ignore transactions with tip lower then ignore price
+				continue
+			}
+
+			sender, err := signer.Sender(tx)
+			if err != nil {
+				return fmt.Errorf("could not get sender of transaction: %s. Error: %w", tx.Hash, err)
+			}
+
+			if sender != blockMiner {
+				blockTxPrices = append(blockTxPrices, tip)
+
+				// if sample number of txs from block is reached,
+				// don't process any more txs
+				if len(blockTxPrices) >= int(g.sampleNumber) {
+					break
+				}
+			}
+		}
+
+		if len(blockTxPrices) == 0 {
+			// either block is empty or all transactions in block are sent by the miner.
+			// in this case add the latests calculated price for sampling
+			blockTxPrices = append(blockTxPrices, lastPrice)
+		}
+
+		// add the block prices to the slice of all prices
+		allPrices = append(allPrices, blockTxPrices...)
+
+		return nil
+	}
+
+	// iterate from current block to previous blocks determined by numOfBlocksToCheck
+	// if chain doesn't have that many blocks, we need to stop the loop (currentBlock.Number() > 0)
+	for i := uint64(0); i < g.numOfBlocksToCheck && currentBlock.Number() > 0; i++ {
+		if err := collectPrices(currentBlock); err != nil {
+			return nil, err
+		}
+
+		currentBlock, found = g.backend.GetBlockByHash(currentBlock.ParentHash(), true)
+		if !found {
+			return nil, fmt.Errorf(couldNotFoundBlockFormat, currentHeader.Number, currentHeader.Hash)
+		}
+	}
+
+	// at least amount of transactions to get
+	minNumOfTx := int(g.numOfBlocksToCheck) * 2
+	if len(allPrices) < minNumOfTx {
+		// collect some more blocks and transactions if not enough transactions were collected
+		for len(allPrices) < minNumOfTx && currentBlock.Number() > 0 {
+			if err := collectPrices(currentBlock); err != nil {
+				return nil, err
+			}
+		}
+	}
+
+	price := lastPrice
+
+	if len(allPrices) > 0 {
+		// sort prices from lowest to highest
+		sort.Sort(bigIntSorted(allPrices))
+		// take the biggest price that is in the configured percentage
+		// by default its 60, so it will take the price on that percentage
+		// of all prices in the array
+		price = allPrices[(len(allPrices)-1)*int(g.pricePercentile)/100]
```

Comment:
- I don't think that this downcasting is generally a good idea. 
Why `pricePercentile` is `uint64` and not `int` (or at least `uint`)?

- Should we validate `pricePercentile`? 

- It is not obvious to me why are you indexing this way (multiplying `pricePercentile` divided by 100 with last index of `allPrices`)? Seems like it can lead to index out of range in some cases.

Reply:
1. Because I don't want percentile to be lower the 0, but ok, I can sanitize it before I set it in the `GasHelper` struct.
2. Answered in 1.
3. Not correct. It will never lead to index ouyt of range because percentile can be in the range of 0-100. As I said, I will sanitize the value provided by config before setting it to `GasHelper`.

Reply:
Fix: https://github.com/0xPolygon/polygon-edge/pull/1629/commits/beeeda2e72ef1a9de256d45961f83db8fe7ff389